providers/common/macs/cmac.c

```diff
@@ -0,0 +1,210 @@
+static OSSL_OP_mac_final_fn cmac_final;
+
+/* local CMAC pkey structure */
+
+/*
+ * TODO(3.0) solve the ENGINE problem
+ *
+ * For the moment, ENGINEs are simply ignored.  A tentative idea is to
+ * pass a property query string instead.  Since EVP_MAC is a new API,
+ * we can really decide whatever we want and make sure that other
+ * libcrypto APIs know how to translate between the two.
+ */
+
+/* typedef EVP_MAC_IMPL */
+struct cmac_data_st {
+    /* tmpcipher and tmpengine are set to NULL after a CMAC_Init call */
+    const EVP_CIPHER *tmpcipher; /* cached CMAC cipher */
+    const ENGINE *tmpengine;     /* cached CMAC cipher engine */
+    /* We need to keep a reference to a fetched cipher */
+    EVP_CIPHER *alloc_cipher;    /* fetched cipher */
+    CMAC_CTX *ctx;
+};
+
+static void *cmac_new(void *unused_provctx)
+{
+    struct cmac_data_st *macctx;
+
+    if ((macctx = OPENSSL_zalloc(sizeof(*macctx))) == NULL
+        || (macctx->ctx = CMAC_CTX_new()) == NULL) {
+        OPENSSL_free(macctx);
+        macctx = NULL;
+    }
+
+    return macctx;
+}
+
+static void cmac_free(void *vmacctx)
+{
+    struct cmac_data_st *macctx = vmacctx;
+
+    if (macctx != NULL) {
+        CMAC_CTX_free(macctx->ctx);
+        EVP_CIPHER_meth_free(macctx->alloc_cipher);
+        OPENSSL_free(macctx);
+    }
+}
+
+static int cmac_copy(void *vdst, void *vsrc)
+{
+    struct cmac_data_st *dst = vdst;
+    struct cmac_data_st *src = vsrc;
+
+    if (!CMAC_CTX_copy(dst->ctx, src->ctx))
+        return 0;
+
+    if (src->alloc_cipher != NULL && !EVP_CIPHER_upref(src->alloc_cipher)) {
+        CMAC_CTX_free(dst->ctx);
+        return 0;
+    }
+
+    dst->tmpengine = src->tmpengine;
+    dst->tmpcipher = src->tmpcipher;
+    dst->alloc_cipher = src->alloc_cipher;
+    return 1;
+}
+
+static size_t cmac_size(void *vmacctx)
+{
+    struct cmac_data_st *macctx = vmacctx;
+
+    return EVP_CIPHER_CTX_block_size(CMAC_CTX_get0_cipher_ctx(macctx->ctx));
+}
+
+static int cmac_init(void *vmacctx)
+{
+    struct cmac_data_st *macctx = vmacctx;
+    int rv = CMAC_Init(macctx->ctx, NULL, 0, macctx->tmpcipher,
+                       (ENGINE *)macctx->tmpengine);
+
+    macctx->tmpcipher = NULL;
+    macctx->tmpengine = NULL;
+
+    return rv;
+}
+
+static int cmac_update(void *vmacctx, const unsigned char *data,
+                       size_t datalen)
+{
+    struct cmac_data_st *macctx = vmacctx;
+
+    return CMAC_Update(macctx->ctx, data, datalen);
+}
+
+static int cmac_final(void *vmacctx, unsigned char *out, size_t *outl,
+                      size_t outsize)
+{
+    struct cmac_data_st *macctx = vmacctx;
+
+    return CMAC_Final(macctx->ctx, out, outl);
+}
+
+static const OSSL_PARAM known_params[] = {
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_ALGORITHM, NULL, 0),
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_CIPHER, NULL, 0),     /* Same as "algorithm" */
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_PROPERTIES, NULL, 0), /* instead of ENGINE */
```

Comment:
Where is `OSSL_MAC_PARAM_PROPERTIES` defined...I don't see it in core_names...or am I going blind?

Reply:
Fixing